### PR TITLE
5.x.x develop listener refactor

### PR DESCRIPTION
**Pull request by TinaTiel**
Large architectural change with Lifecycle listeners and immutable builders. Summary:

- Communicator has a builder
- Communicator has its own lifecycle listener
- Remote has a builder
- Remote has its own lifecycle listeners
- Lifecycle listeners all have their own builders as well, using callback semantics like before (since that was well loved)
- "debug" boolean has been removed from both Communicator and Remote. Old methods using it have been marked deprecated -- we could decide to remove them altogether in 2.X.X
- Builders always return Composite listeners; these take many listeners and execute them in order
- By default Lifecycle Builders include a LoggingListener variant; this can be disabled or replaced if desired
- All Integration and Unit tests thoroughly reviewed...
  - There are a few that would be good to have discussion on, not sure their point
  - Many "unit" tests were actually integration (or E2E) tests; they've been moved to integration test suites
  - Many former Integration tests are now unit tests; what they tested is now managed by the listeners

What I'd like to hear feedback on especially:

- any lifecycle callbacks that are missing?
- the purpose of the "isFailed" boolean in the Controller?
- do we keep the old constructors as deprecated, or remove them?
- any commented out tests -- remove them? Or missing some kind of coverage?

Want to note -- there is a LOT of commented out code in this PR. 
I've left that in there for context so we can still look at it / decipher some original intent -- but once all seems well in the PR, I think we could clean it up now.

### CODE REVIEW

**Comment by TinaTiel** [comment]
Merged commits from #46

File: build.gradle [hunk 1]
```diff
@@ -22,6 +22,7 @@ java.sourceCompatibility = JavaVersion.VERSION_1_8
 repositories {
     mavenLocal()
     mavenCentral()
+    jcenter()
```
Comment (ChristopheCVB) [inline build.gradle:25]: JCenter repo is deprecated, please remove it

**Review by ChristopheCVB — APPROVED** [review]
Aside the jscenter repo, everything looks good

**Comment by TinaTiel** [comment]
> Aside the jscenter repo, everything looks good

Mockito requires jcenter according to their website: https://site.mockito.org/
I"m going to verify it can run without...

**Comment by ChristopheCVB** [comment]
> > Aside the jscenter repo, everything looks good
> 
> Mockito requires jcenter according to their website: https://site.mockito.org/
> I"m going to verify it can run without...

It's also available on Maven Central https://mvnrepository.com/artifact/org.mockito/mockito-core

**Comment by TinaTiel** [comment]
Thanks Christophe!
I'm not very experienced with Gradle, greatly appreciate your guidance on this. 
I'll go ahead and merge, can double-confirm on my end too all looks well.

**Pull request merged**